p2p/src/peers/mod.rs

```diff
@@ -213,39 +214,52 @@ impl Peers {
     }
 
     /// Get a random socket address from the peers list
-    pub fn get_random(&self) -> Result<Option<SocketAddr>, failure::Error> {
-        let bucket = match (self.new_bucket.is_empty(), self.tried_bucket.is_empty()) {
-            (true, true) => return Ok(None),
-            (true, false) => &self.tried_bucket,
-            (false, true) => &self.new_bucket,
-            (false, false) => {
-                if thread_rng().gen() {
-                    &self.tried_bucket
-                } else {
-                    &self.new_bucket
-                }
-            }
-        };
+    pub fn get_random_peers(&self, n: usize) -> Result<Vec<SocketAddr>, failure::Error> {
+        let mut rng = rand::thread_rng();
 
-        // Random index with range [0, len) of the peers vector
-        let index = thread_rng().gen_range(0, bucket.len());
+        let max_tried = self.tried_bucket.len();
+        let max_new = self.new_bucket.len();
 
-        Ok(bucket.values().nth(index).map(|v| v.address.to_owned()))
-    }
+        let n_peers = cmp::min(max_tried + max_new, n);
 
-    /// Get a random socket address from the new peers list
-    pub fn get_new_random(&self) -> Option<(u16, SocketAddr)> {
-        if self.new_bucket.is_empty() {
-            return None;
+        let mut v_peers: Vec<SocketAddr> = vec![];
+
+        if n_peers < 1 {
+            return Ok(v_peers);
         }
 
-        // Random index with range [0, len) of the peers vector
-        let index = thread_rng().gen_range(0, self.new_bucket.len());
+        let min_new = cmp::max(0, n_peers as i32 - max_tried as i32) as usize;
+        let index = if min_new == n_peers {
+            min_new
+        } else {
+            rng.gen_range(min_new, n_peers)
+        };
+        let index = cmp::min(max_new, index);
+        v_peers.extend(
+            self.new_bucket
+                .values()
+                .map(|p| p.address)
+                .choose_multiple(&mut rng, index),
+        );
+        v_peers.extend(
+            self.tried_bucket
+                .values()
+                .map(|p| p.address)
+                .choose_multiple(&mut rng, n_peers - index),
+        );
 
-        self.new_bucket
-            .iter()
-            .nth(index)
-            .map(|(k, v)| (*k, v.address.to_owned()))
+        Ok(v_peers)
+    }
+
+    /// Get a random socket address from the new peers list
+    pub fn get_new_random_peer(&self) -> Option<(u16, SocketAddr)> {
+        let mut rng = rand::thread_rng();
+        let key = self.new_bucket.keys().choose(&mut rng);
```

Comment:
You can use `.iter()` to get a key with its value:

```suggestion
        self.new_bucket.iter().choose(&mut rng).map(|k, v| (*k, v.address))
```